### ISSUE TITLE
USB redirection

### ISSUE THREAD

**Issue by imsonica** [1Remote/1Remote#113]
Hello,
is it possible to add the feature to redirect the USB devices to the remote host?
I've selected the "port" flag on the options tab of the connection but I'm not able to see the "Select device to use" button as in Microsoft RD client

![2021-04-22 20_59_26-7976 - cg007976 - Remote Desktop Connection](https://user-images.githubusercontent.com/64767787/115771281-0ba60a00-a3ae-11eb-9119-69773fe1c276.png)

Thank you so much, your software is working very well :)

A

**Comment by VShawn** [comment]
wooah sorry I didn't realize it doesn't support usb redirection, I will find out the solution.


---

it seems that the mstsc librariy offering by Microsoft, did not offer a special option for usb redirection, it must be hiding somewhere.

![image](https://user-images.githubusercontent.com/10143738/115799699-9e28c680-a40b-11eb-93cf-8aa5414fc94a.png)

**Comment by imsonica** [comment]
for your info it is called "RemoteFX USB Device Redirection" and has to be enabled in client and host via gpedit
![2021-04-23 07_43_33-USB redirection · Issue #113 · VShawn_PRemoteM](https://user-images.githubusercontent.com/64767787/115824193-e1813600-a407-11eb-85a4-08239dc16164.png)

**Comment by VShawn** [comment]
I developed a beta version and can redirect my USB flash drive. Now it only need to check the "Disk Drives" then the USB flash drive will be redirected. Since I only have a USB flash drive on hand now, I cannot test whether other USB devices are working properly.

If you have other USB devices, please help me to test.
[PRemoteM 0.5.10.3-beta.zip](https://github.com/VShawn/PRemoteM/files/6363396/PRemoteM.0.5.10.3-beta.zip)


![image](https://user-images.githubusercontent.com/10143738/115833961-3bedb700-a447-11eb-9102-8de4eb47e623.png)

And I am still looking for how to show this “devices button” on the connection bar : (, I cannot find anything about how to show this button.

https://docs.microsoft.com/en-us/windows/win32/termserv/imstscaxevents-ondevicesbuttonpressed

**Comment by nicobo01** [comment]
Hello @VShawn ,
Your application is detected as ransomware.

**Comment by VShawn** [comment]
@nicobo01 
err,sorry but I don't understand why your anti-virus software misreported.My package was compiled directly through msbuild, and it shouldn't be misreported(IMO). I don't know the knowledge of anti-virus, so I can't analyze the reason.

BTW, I use windows defender only, and it gets along well with my application.

**Comment by VShawn** [comment]
@nicobo01 Are all the packages(https://github.com/VShawn/PRemoteM/releases) are detected as ransomware,
or only this version(PRemoteM 0.5.10.3-beta.zip) is reported as a virus?

in this version I added a subscription to the USB device unplugging event, it may be that this place caused the virus misreported.

**Comment by nicobo01** [comment]
Hello @VShawn 

My antivirus is Trend Micro Apex-One. This detects PRemoteM.exe as a potential ransomware.
I did not test with another version of the software.
The offending one is the PRemoteM.0.5.10.3-beta version. 

![2021-04-23_18-57-44](https://user-images.githubusercontent.com/83074498/115906830-597f4880-a468-11eb-897a-67976eaef536.jpg)
,

**Comment by VShawn** [comment]
@nicobo01 thank you, I will download it for a free trial latter.

**Comment by imsonica** [comment]
> I developed a beta version and can redirect my USB flash drive. Now it only need to check the "Disk Drives" then the USB flash drive will be redirected. Since I only have a USB flash drive on hand now, I cannot test whether other USB devices(like a camera [mRemoteNG/mRemoteNG#1951](https://github.com/mRemoteNG/mRemoteNG/issues/1951)) are working properly.
> 
> If you have other USB devices, please help me to test.
> [PRemoteM 0.5.10.3-beta.zip](https://github.com/VShawn/PRemoteM/files/6363396/PRemoteM.0.5.10.3-beta.zip)
> 
> ![image](https://user-images.githubusercontent.com/10143738/115833961-3bedb700-a447-11eb-9102-8de4eb47e623.png)
> 
> And I am still looking for how to show this “devices button” on the connection bar : (, I cannot find anything about how to show this button.
> 
> https://docs.microsoft.com/en-us/windows/win32/termserv/imstscaxevents-ondevicesbuttonpressed

The actual beta doesnt do the job, usb devices are not redirected.

Ciao

**Comment by VShawn** [comment]
@imsonica 

> The actual beta doesnt do the job, usb devices are not redirected.

usb devices redirect only in tab view, rdp with full screen will not work, it is a bug i fixed 11 hours ago  https://github.com/VShawn/PRemoteM/commit/771b0f6051e770189b4338abe6c6af9a91a1de44. do you test in full-screen mode? and Is your USB device a USB flash drive?

**Comment by imsonica** [comment]
> @imsonica
> 
> > The actual beta doesnt do the job, usb devices are not redirected.
> 
> usb devices redirect only in tab view, rdp with full screen will not work, it is a bug i fixed 11 hours ago [771b0f6](https://github.com/VShawn/PRemoteM/commit/771b0f6051e770189b4338abe6c6af9a91a1de44). do you test in full-screen mode? and Is your USB device a USB flash drive?

I'll try another test, I'm not sure I can redirect usb since my company policies. I tested and I would like to redirect usb HID devices like spacemouce.
I did the test in full screen, I'll try again as you said.

**Comment by VShawn** [comment]
Thank you for your help, but according to your description, I feel that my previous work cannot fit your needs. And what's more regrettable is that when I don't have a spacemouce, I can't debug this function at all.

**Comment by imsonica** [comment]
> Thank you for your help, but according to your description, I feel that my previous work cannot fit your needs. And what's more regrettable is that when I don't have a spacemouce, I can't debug this function at all.

In the list of RemoteFX also webcams or usb headsets appear, you could debug using these kind of devices.

**Comment by majkinetor** [comment]
I can't see a difference - v0.5.10.2 already shows USB drive as remote drive. I can't see any new local drive in both versions.

**Comment by VShawn** [comment]
@majkinetor 

> I can't see a difference - v0.5.10.2 already shows USB drive as remote drive. I can't see any new local drive in both versions.

The older version cannot update the device redirection when you insert or remove the USB flash drive.

see:
![1](https://user-images.githubusercontent.com/10143738/116057039-fdf3cb80-a6b0-11eb-9597-493b93a54043.gif)

**Comment by majkinetor** [comment]
No, it doesn't work for me. It shows under 'Redirected drives', but that worked on old version. Tried 2 different USB drives and 2 different Windows remote OS.

**Comment by VShawn** [comment]
@itagagaki Thanks for your log. It looks like a error caused by character encoding or reflection from the log message. I delete some plugin dll, it may works.


[PRemoteM.0.5.10.3-beta2.zip](https://github.com/VShawn/PRemoteM/files/6458726/PRemoteM.0.5.10.3-beta2.zip)

**Comment by itagagaki** [comment]
Beta2 didn't work either but I found out that one of the reasons it didn't work was because of the zone identifiers, so I unblocked them and tried to run it, and got a Chinese error dialog. I do not understand what it says..
![image](https://user-images.githubusercontent.com/71218184/117810653-a3569400-b29a-11eb-8e60-287ffd786c46.png)

**Comment by VShawn** [comment]
@itagagaki  it said It does not have permission to read the database, or the database is already occupied.  

The error message was decided to be Chinese by environment judgement, It seems that my program is still not able to adapt to the Japanese system environment. 😞

**Comment by itagagaki** [comment]
Okay, I was able to launch 0.5.10.3 beta.
I'll finally get to the point. How can I test the redirection of cameras and other USB devices?
I checked all of the "Resources" for now, but unfortunately the remote PC doesn't find the camera.
![image](https://user-images.githubusercontent.com/71218184/117816075-17943600-b2a1-11eb-8d52-68ba0cac3cad.png)

**Comment by VShawn** [comment]
@itagagaki  this means usb redirection not works for usb cameras.  It seems that I have to buy a usb camera📷  to debug this.  

BTW did you test usb flash drive? did it works?

**Comment by majkinetor** [comment]
>  I have to buy a usb camera📷 to debug this.

I believe modern smart phones are the same with their camera mode of usb transfer.

**Comment by itagagaki** [comment]
@VShawn usb flash drive works but it works also on 0.5.10.2

**Comment by majkinetor** [comment]
No, that is not it, previously it showed as remote drive (from the host) while it should show as local drive.

**Comment by itagagaki** [comment]
Ah, I see. Yes, it showed up as a remote drive, but not as a local drive.

**Comment by Xp96** [comment]
> Desenvolvi uma versão beta e posso redirecionar minha unidade flash USB. Agora basta verificar as "Unidades de disco" para que a unidade flash USB seja redirecionada. Como agora só tenho uma unidade flash USB disponível, não posso testar se outros dispositivos USB (como uma câmera [mRemoteNG / mRemoteNG # 1951](https://github.com/mRemoteNG/mRemoteNG/issues/1951) ) estão funcionando corretamente.
> 
> Se você tiver outros dispositivos USB, por favor me ajude a testar. [PRemoteM 0.5.10.3-beta.zip](https://github.com/VShawn/PRemoteM/files/6363396/PRemoteM.0.5.10.3-beta.zip)
> 
> ![imagem](https://user-images.githubusercontent.com/10143738/115833961-3bedb700-a447-11eb-9102-8de4eb47e623.png)
> 
> E ainda estou procurando como mostrar este “botão de dispositivos” na barra de conexão: (, não consigo encontrar nada sobre como mostrar este botão.
> 
> https://docs.microsoft.com/en-us/windows/win32/termserv/imstscaxevents-ondevicesbuttonpressed


I think this may help sorry but i didn't find it in english
https://www.ti-enxame.com/pt/window/como-habilitar-o-redirecionamento-usb-no-windows-10/1056392685/

**Comment by Xp96** [comment]
> I developed a beta version and can redirect my USB flash drive. Now it only need to check the "Disk Drives" then the USB flash drive will be redirected. Since I only have a USB flash drive on hand now, I cannot test whether other USB devices(like a camera [mRemoteNG/mRemoteNG#1951](https://github.com/mRemoteNG/mRemoteNG/issues/1951)) are working properly.
> 
> If you have other USB devices, please help me to test. [PRemoteM 0.5.10.3-beta.zip](https://github.com/VShawn/PRemoteM/files/6363396/PRemoteM.0.5.10.3-beta.zip)
> 
> ![image](https://user-images.githubusercontent.com/10143738/115833961-3bedb700-a447-11eb-9102-8de4eb47e623.png)
> 
> And I am still looking for how to show this “devices button” on the connection bar : (, I cannot find anything about how to show this button.
> 
> https://docs.microsoft.com/en-us/windows/win32/termserv/imstscaxevents-ondevicesbuttonpressed

I think this may help sorry but i didn't find it in english
https://www.ti-enxame.com/pt/window/como-habilitar-o-redirecionamento-usb-no-windows-10/1056392685/

**Comment by brendan-pike** [comment]
I have been trying to get a USB footpedal to work over RDP with no luck.  It will work to the same server when I use a Linux client (Remmina) but not on my Windows 10.  I have enabled all of the Advantage / Resources but the foot pedal does not show up.  Is there anything I can do it get this working?

**Comment by VShawn** [comment]
@brendan-pike  We lost usb redirection after port to .NET6. Still dont know how to bring it back.

You can try https://github.com/1Remote/PRemoteM/releases/tag/0.6.2.1, Usb disk redirection works on this version.

**Comment by brendan-pike** [comment]
Thanks, no joy with that version either sadly.  I should of mentioned that the official windows RDP client doesn't work with my footpedal either. So it may not be specific to this app, was just hoping to find a windows one that worked like Llinux can.

**Comment by VShawn** [comment]
Hi guys, I still working on this ticke.

---

> We lost usb redirection when port to .NET6 after 0.7.X.

The usb disk functional failure because of `pointer size in 64 bit` has been successfully repaired today. It will be merged into PRemoteM after I've been testing it for a while.

---


FYI


Different between RDP redirection and RemoteFX one:

![image](https://user-images.githubusercontent.com/10143738/208287975-5f968669-1726-4f97-85de-5a3d8cc7e82b.png)


USB Redirection method for different devices: 

![image](https://user-images.githubusercontent.com/10143738/208287625-31695ea1-b408-4b29-9a53-74c943176306.png)

**Comment by itagagaki** [comment]
It's been a while. I saw your new comment and tried again 0.7.1.7 and the Microsoft Store version. The local webcam and headset were not recognized by the remote; the USB drive is recognized. I also tried changing the group policy but no luck.

**Comment by VShawn** [comment]
> It's been a while. I saw your new comment and tried again 0.7.1.7 and the Microsoft Store version. The local webcam and headset were not recognized by the remote; the USB drive is recognized. I also tried changing the group policy but no luck.

Hey I've not make a releass for this yet. Besides the new commit fix the usb disk plug in event redirect. usb camera redirect is still **WIP**.

BTW if you wannt try usb disk redirect you can get nightly build form https://github.com/1Remote/PRemoteM/actions/runs/3728969726. Note you might need enable the option:

![image](https://user-images.githubusercontent.com/10143738/208360969-a370e07a-93f0-4b4e-91cf-f30fe3d57335.png)

**Comment by itagagaki** [comment]
I found that by setting mstsc.exe mode to enabled, a USB devices connected to the client PC (e.g. webcam) can be used on the remote PC !
However, that does not allow the remote's display size to be the client's window size. Nor can the tab function be used. Then it is the same as Microsoft's RDP client, so there is no point in using 1Remote.
Is this unavoidable?

**Comment by VShawn** [comment]
@itagagaki 
Nope, technically I have made camera available via tab mode. I testing it all the time when I stay home with a covid19 fever last week and finally it works.


![image](https://user-images.githubusercontent.com/10143738/210024558-aa7526ba-b1f3-4b7d-a2ec-eb6a09f9b338.png)

**Comment by majkinetor** [comment]
> Remmina

@brendan-pike maybe you can use Remmina for such connections via custom protocol, if it can be controled via CLI?

**Comment by VShawn** [comment]
I haven't tested the gamepad yet, I'll next step buy a Xbox controller to see if it can be redirected:)

**Comment by itagagaki** [comment]
Hi,
I built and ran the latest code, but on my end, when mstsc.exe mode is disabled, the webcam did not appear on the remote side.
This may be unrelated, but on my client PC (Windows 10 Pro 21H2), the camera was in Device Manager under "Imaging Devices", not "Cameras".

While trying to try further, an exception was thrown.
'attr' was null at LauncherWindowView.xaml.cs Line 387.

**Comment by VShawn** [comment]
@imsonica sorry I haven't committed this part of the code yet:joy:

**Comment by majkinetor** [comment]
What is the status of this @VShawn , can it be closed or not ?

**Comment by VShawn** [comment]
@majkinetor 

the tasks:

- [x] USB disk 
- [x] USB camera
- [ ] USB gamepad based on RemoteFX

Since I have no idear for RemoteFX now. We can close this ticket and create a new [Limbo](https://github.com/1Remote/1Remote/milestone/2) ticket for `USB gamepad based on RemoteFX`.

**Comment by itagagaki** [comment]
> > webcam did not appear on the remote side.
> 
> @imsonica sorry I haven't committed this part of the code yet😂

Is this not reflected in the release binary yet?
I tried 0.7.2.8 but could not redirect the USB webcam.

**Comment by VShawn** [comment]
no, camera no working on 0.7.2.8

you can find it on 1Remote version.

![image](https://github.com/1Remote/1Remote/assets/10143738/f00d709b-7cf6-4fe4-848f-f953409c7276)

**Comment by itagagaki** [comment]
@VShawn Where is the 1Remote version?
I downloaded from below, but I can only find PRemote.exe but 1Remote in it.
https://github.com/1Remote/1Remote/releases/tag/0.7.2.8

**Comment by VShawn** [comment]
@itagagaki 
here: https://1remote.org/download/

it still in public beta test now.

**Comment by itagagaki** [comment]
@VShawn 
So you mean here?
https://github.com/1Remote/1Remote/releases/tag/Nightly

I'll wait for stable release.
Thank you.
